src/AdoNet/Orleans.Persistence.AdoNet/Options/AdoNetGrainStorageOptions.cs

```diff
@@ -0,0 +1,138 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.Extensions.Options;
+using Newtonsoft.Json;
+using Orleans.Persistence.AdoNet.Storage;
+using Orleans.Runtime;
+using Orleans.Runtime.Configuration;
+
+namespace Orleans.Hosting
+{
+    /// <summary>
+    /// Options for AdonetGrainStorage
+    /// </summary>
+    public class AdoNetGrainStorageOptions
+    {
+        /// <summary>
+        /// Connection string for AdoNet storage.
+        /// </summary>
+        public string ConnectionString { get; set; }
+
+        /// <summary>
+        /// Stage of silo lifecycle where storage should be initialized.  Storage must be initialzed prior to use.
+        /// </summary>
+        public int InitStage { get; set; } = DEFAULT_INIT_STAGE;
+        /// <summary>
+        /// Default init stage in silo lifecycle.
+        /// </summary>
+        public const int DEFAULT_INIT_STAGE = SiloLifecycleStage.ApplicationServices;
+
+        /// <summary>
+        /// The default ADO.NET invariant used for storage if none is given. 
+        /// </summary>
+        public const string DEFAULT_ADONET_INVARIANT = AdoNetInvariants.InvariantNameSqlServer;
+        /// <summary>
+        /// The invariant name for storage.
+        /// </summary>
+        public string AdoInvariant { get; set; } = DEFAULT_ADONET_INVARIANT;
+
+        #region json serialization related settings
+        /// <summary>
+        /// Whether storage string payload should be formatted in JSON.
+        /// <remarks>If neither <see cref="UseJsonFormat"/> nor <see cref="UseXmlFormat"/> is set to true, then BinaryFormatSerializer will be configured to format storage string payload.</remarks>
+        /// </summary>
+        public bool UseJsonFormat { get; set; }
+        public bool UseFullAssemblyNames { get; set; }
+        public bool IndentJson { get; set; }
+        public TypeNameHandling? TypeNameHandling { get; set; }
+        #endregion
+        /// <summary>
+        /// Whether storage string payload should be formatted in Xml.
+        /// <remarks>If neither <see cref="UseJsonFormat"/> nor <see cref="UseXmlFormat"/> is set to true, then BinaryFormatSerializer will be configured to format storage string payload.</remarks>
+        /// </summary>
+        public bool UseXmlFormat { get; set; }
```

Comment:
Should these be an enum? I'd prefer if we could somehow remove them, but maybe not for 2.0

Reply:
We can revamp the serializer configuration after 2.0 rc, including improve configuration, make it possible to use other serializer. Just as @veikkoeeva mentioned 

But without doing all that, just make this an enum doesn't bring much benefits IMO.

Reply:
I'm ok with leaving it for now

Reply:
Ideally these wouldn't be needed. As mentioned elsewhere, the idea is to allow any format to the database through the canonical interfaces and then just note which "type slot" to fit them in the database (binary, XML and JSON) where binary really swallos anything, but JSON or XML could be preferred for other reasons (e.g. in-storage processing).

Reply:
@veikkoeeva you haven't answered my question above. How is serializer picker feature suppose to work? I copied my original question below: 

Also, I have some question about this (de)serializer feature. Those (de)serializer are wrapped into their own canonical interfaces, and further put into a `DefaultRelationalStoragePicker`. This confuses me because it seems like only one (de)serializer is allowed. So why do we have a picker? And also there's a ` public IStorageSerializationPicker StorageSerializationPicker { get; set; }` on `AdoNetStorageProvider`, which seems to provide a way to modify the picker dynamically in runtime. I guess I need more information on how this feature are supposed to work.

Reply:
You are the last person touched the code, if you know background on this, much appreciation on sharing it.

Reply:
Sorry, @xiazen I just missed it. It seems to be I need to be excessively pinged occassionally. :)

My original idea was that:
- Any (de)serializer can be wrapped into a canonical interface and loaded and handed over from elsewhere. That's the wrapping part. One complication is that there might be business or technical reasons to favour one kind of a storage layout over other (i.e. binary, JSON or XML), so it's indicated by tagging.

- The name `IStorageSerializationPicker` is not particulary good one. The idea is that this contains a rule to choose a (de)serializer with the parameters you see there. Since it's not possible to write a general rule to choose a (de)serializer, the idea is that by default one (de)serializer is instantiated with a rule to pick it. But it would allow one to use multiple (de)serializers with quite many parameters if the users could implement `IStorageSerializationPicker` by self.

There is one test exercising this idea by storing data (serializing) in one format and then reading (deserializing) it in other format. Practically switching the format. This could be used to:
1. Switch from Orleans binary to JSON, for instance.
2. But also implement custom logic to otherwise reshape data being read from the database (or being inserted to the database). This should make refactoring easier, or fixing "data bugs" easier.

In real deployment this could work so that in "deployment number n1" one deploys some customer transformation logic that checks the data to determine if it's in some format (say, some fields are present, indicating old format, so reshaped into new format consistent with new state shape) and once all is converted, during the next round of deployments just remove the customer refactoring code. The user has, naturally, also the option to do this as in-storage processing too.

But I'm not entirely sure how this should be exposed, what should be the naming of the classes be, the "surface" could be easier too (e.g. separate the reshaping logic into a lambda or some other interface from the picker logic).

Thinking in more general terms, keeping things consistent throughout Orleans, this is basically pipes-and-filters, VET(R)O, interceptors etc. and might be something to consider applied later elsewhere too (e.g. when deploying new grains, if not every can be handled by deactivating grains in old format and routing calls to grains accepting new format).

Reply:
Originally I had an idea to inject these to these to the storage provider, but there were problems. I also tried a bit to "detach" the tests from the quite tight coupling the then system and work a bit on logic on how to ensure the system is ready for testing (initialize database etc.) and then actually to run the tests (so that they look OK also in the explorer).

I have ideas that should clean up the DB initialization logic considerable and introduce also construction injection with XUnit.NET to the tests making it cleaner still (I mentioned this in https://github.com/dotnet/orleans/issues/3969).

Maybe these help. If not, do ping me. I believe I'm off to bed soon, but I'll be reading again 8-9 hours before leaving to work. It's also end of sprint tomorrow too, so I'm likely more around in chat etc. too.

Reply:
Saw it :) in case you worry I didn't see this. Thanks for providing the details on how the feature was supposed to work. I meant to bring those to the issue I created too. But I kind of forgot to do it in the process. RC is a bit crazy. Will do it now.